### PR TITLE
UserService_CreateUser

### PR DESCRIPTION
**Pull request by denis-kriv**
Add main logic for add user and tests for it

### CODE REVIEW

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+using UserService.Database;
+using UserService.Database.Entities;
+using UserService.Interfaces;
+using UserService.Repositories;
+using UserService.Repositories.Interfaces;
+using UserService.RestRequests;
+
+namespace UserService.Commands
+{
+    public class RegisterCommand : IRegisterCommand
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: UserCreateCommand

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        private readonly IValidator<RegisterRequest> validator;
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: UserCreateRequest

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        public RegisterCommand([FromServices] IValidator<RegisterRequest> validator, [FromServices] IUserRepository repository)
+        {
+            this.validator = validator;
+            this.repository = repository;
+        }
+
+        public bool Execute(RegisterRequest request)
+        {
+            try
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: remove try

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            }
+            catch (Exception)
+            {
+                throw new Exception("Error in fields.");
+            } 
+            var user = new User
+            {
+                Id = Guid.NewGuid(),
+                Email = request.Email,
+                FirstName = request.Email,
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: ???????

**Review by DarkFelas — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs (hunk 4 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: fixed

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs (hunk 3 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: fixed

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs (hunk 2 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: fixed

File: DigitalOffice/Services/UserService/Commands/RegisterCommand.cs (hunk 1 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/RegisterCommand.cs]: fixed

File: DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                Email = request.Email,
+                FirstName = request.FirstName,
+                LastName = request.LastName,
+                MiddleName = request.MiddleName,
+                Status = request.Status,
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes(request.Password))).ToString(),
+                AvatarFileId = null,
+                IsActive = true,
+                CertificatesFilesIds = new Collection<UserCertificateFile>(),
+                PositionsIds = new Collection<UserPosition>(),
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: UserService in develop branch has not UserPosition entity, so build failed after merge with develop . In your branch you have old versions of Entities. Pls fix it.

Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: fixed

**Review by lalagvana — APPROVED** [review]

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    class UserRepositoryTests
+    { 
+        private IUserRepository repository;
+        private DbContextOptions<UserServiceDbContext> dbOptionsRegisterUser;
+
+        [SetUp]
+        public void Initialize()
+        {     
+            dbOptionsRegisterUser = new DbContextOptionsBuilder<UserServiceDbContext>()
+                .UseInMemoryDatabase(databaseName: "RegsterUserTest")
```
Comment (lalagvana) [inline DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Fix mistake to "RegisterUserTest", please.

Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

File: DigitalOffice/Services/UserService/Repositories/UserRepository.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        public bool UserCreate(User user)
+        {
+            if (dbContext.Users.Any(users => user.Email == users.Email))
+            {
+                throw new Exception("Email is already taken.");
+            }
+            dbContext.Users.Add(user);
+            dbContext.SaveChanges();
+            return true;
+        }
```
Comment (djino97) [inline DigitalOffice/Services/UserService/Repositories/UserRepository.cs]: where are empty lines?

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            var request = new UserCreateRequest
+            {
+                FirstName = "Example",
+                LastName = "Example",
+                MiddleName = "Example",
+                Email = "Example@gmail.com",
+                Status = "Example",
+                Password = "Example"
+            };
+            Assert.Throws<ValidationException>(() => command.Execute(request));
```
Comment (djino97) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: where are empty lines?

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            var request = new UserCreateRequest
+            {
+                FirstName = "Example",
+                LastName = "Example",
+                MiddleName = "Example",
+                Email = "Example@gmail.com",
+                Status = "Example",
+                Password = "Example"
+            };
+            Assert.Throws<Exception>(() => command.Execute(request), "Email is already taken.");
```
Comment (djino97) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: where are empty lines?

File: DigitalOffice/Services/UserService/Startup.cs [hunk 10]
```diff
@@ -25,6 +32,9 @@ public void ConfigureServices(IServiceCollection services)
             });
 
             services.AddControllers();
+            services.AddTransient<IUserCreateCommand, UserCreateCommand>();
+            services.AddTransient<IValidator<UserCreateRequest>, UserCreateRequestValidator>();
+            services.AddTransient<IUserRepository, UserRepository>();
```
Comment (djino97) [inline DigitalOffice/Services/UserService/Startup.cs]: move it into new method.

**Review by djino97 — APPROVED** [review]

File: DigitalOffice/Services/UserService/Startup.cs (hunk 10 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Startup.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 9 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 8 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Services/UserService/Repositories/UserRepository.cs (hunk 7 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Repositories/UserRepository.cs]: fixed

File: DigitalOffice/Services/UserService/Interfaces/IUserCreateCommand.cs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿using UserService.RestRequests;
+
+namespace UserService.Interfaces
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Interfaces/IUserCreateCommand.cs]: Please, move Interfaces directory to Commands directory.

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 6 above)
Comment (MihailKulikov) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Fix failed. There is still typo.

File: DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+                Email = request.Email,
+                FirstName = request.FirstName,
+                LastName = request.LastName,
+                MiddleName = request.MiddleName,
+                Status = request.Status,
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes(request.Password))).ToString(),
+                AvatarFileId = null,
+                IsActive = true,
+                CertificatesFilesIds = new Collection<UserCertificateFile>(),
+                AchievementsIds = new Collection<UserAchievement>()
```
Comment (Inzilbeth) [inline DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs:44]: Shouldn't there be a null check for some of the parameters that are not necessary? Middle name, for example, might be missing in the request.

File: DigitalOffice/Services/UserService/Interfaces/IUserCreateCommand.cs (hunk 11 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Interfaces/IUserCreateCommand.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/RegisterRequestValidatorTests.cs [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+                MiddleName = "Example",
+                Email = "Example@gmail.com",
+                Status = "Example",
+                Password = "Example"
+            };
+            Assert.Throws<ValidationException>(() => validator.ValidateAndThrow(request), "First name is empty.");
+        }
+
+        [Test]
+        public void FirstNameContainsNumbers()
```
Comment (Inzilbeth) [inline DigitalOffice/Tests/UserServiceTests/ValidatorsTests/RegisterRequestValidatorTests.cs]: FirstNameContainsNumbers... So how should the programm react? I'd call this and all the test methods like "ShouldDoSomethingIfSomethingHappened"

File: DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (Inzilbeth) [inline DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: Shouldn't there be a null check for unnecessary fields that might be missing in request?

Comment (Inzilbeth) [reply DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: Nope, this is not needed

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/RegisterRequestValidatorTests.cs (hunk 13 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/ValidatorsTests/RegisterRequestValidatorTests.cs]: fixed

File: DigitalOffice/Services/UserService/Startup.cs [hunk 15]
```diff
@@ -25,6 +32,15 @@ public void ConfigureServices(IServiceCollection services)
             });
 
             services.AddControllers();
+
+            CreateTransientObjects(services);
+        }
+
+        private void CreateTransientObjects(IServiceCollection services)
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Startup.cs]: Not really good, create separate methods for each type of objects e.g. repositories, commands ....

File: DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            this.validator = validator;
+            this.repository = repository;
+        }
+
+        public bool Execute(UserCreateRequest request)
+        {
+            validator.ValidateAndThrow(request);
+
+            var user = new User
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: Create mapper

File: DigitalOffice/Services/UserService/Repositories/UserRepository.cs [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly UserServiceDbContext dbContext;
+
+        public UserRepository(UserServiceDbContext dbContext)
+        {
+            this.dbContext = dbContext;
+        }
+
+        public UserRepository()
+        {
+            dbContext = new UserServiceDbContext(new DbContextOptions<UserServiceDbContext>());
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Repositories/UserRepository.cs]: Use DI

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+using UserService.Interfaces;
+using UserService.Repositories.Interfaces;
+using UserService.RestRequests;
+
+namespace UserServiceUnitTests.CommandsTests
+{
+    class UserCreateCommandTests
+    {
+        private Mock<IValidator<UserCreateRequest>> validatorMock;
+        private Mock<ValidationResult> validatorResultMock;
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: No need, remove

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+                .Setup(x => x.Validate(It.IsAny<IValidationContext>()))
+                .Returns(validatorResultMock.Object);
+
+            command = new UserCreateCommand(validatorMock.Object,repositoryMock.Object);
+        }
+
+        [Test]
+        public void ShouldThrowExceptionWhenUserDataIsInvalid()
+        {
+            validatorResultMock
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: Remove

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+                Password = "Example"
+            };
+
+            Assert.Throws<ValidationException>(() => command.Execute(request));
+        }
+
+        [Test]
+        public void ShouldThrowExceptionWhenEmailIsAlreadyTaken()
+        {
+            validatorResultMock
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: Remove

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+                Password = "Example"
+            };
+
+            Assert.Throws<Exception>(() => command.Execute(request), "Email is already taken.");
+        }
+
+        [Test]
+        public void ShouldCreateUserWhenUserDataIsValid()
+        {
+            validatorResultMock
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: Remove

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+{
+    class UserRepositoryTests
+    { 
+        private IUserRepository repository;
+        private DbContextOptions<UserServiceDbContext> dbOptionsCreateUser;
+
+        [SetUp]
+        public void Initialize()
+        {     
+            dbOptionsCreateUser = new DbContextOptionsBuilder<UserServiceDbContext>()
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Read do-backend-wiki

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+                Status = "Example",
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes("Example"))).ToString(),
+                AvatarFileId = null,
+                IsActive = true,
+                CertificatesFilesIds = new Collection<UserCertificateFile>(),
+                AchievementsIds = new Collection<UserAchievement>()
+            };
+            using var dbContext = new UserServiceDbContext(dbOptionsCreateUser);
+            repository = new UserRepository(dbContext);
+            Assert.DoesNotThrow(() => repository.UserCreate(user));
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Assert.True

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+                Status = "Example",
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes("Example"))).ToString(),
+                AvatarFileId = null,
+                IsActive = true,
+                CertificatesFilesIds = new Collection<UserCertificateFile>(),
+                AchievementsIds = new Collection<UserAchievement>()
+            };
+            using var dbContext = new UserServiceDbContext(dbOptionsCreateUser);
+            repository = new UserRepository(dbContext);
+            repository.UserCreate(user);
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Remove

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 22 above)
Comment (DarkFelas) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Add test user to dbContext here

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+            var request = new UserCreateRequest
+            {
+                FirstName = "",
+                LastName = "Example",
+                MiddleName = "Example",
+                Email = "Example@gmail.com",
+                Status = "Example",
+                Password = "Example"
+            };
+            Assert.Throws<ValidationException>(() => validator.ValidateAndThrow(request), "First name is empty.");
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs]: Read do-backend-wiki and refactor all tests

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+                FirstName = "Example",
+                LastName = "Example",
+                MiddleName = "Example",
+                Email = "Example@gmail.com",
+                Status = "Example",
+                Password = ""
+            };
+            Assert.Throws<ValidationException>(() => validator.ValidateAndThrow(request), "Password is empty.");
+        }
+
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs]: Remove one line

**Review by DarkFelas — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/UserService/Startup.cs (hunk 15 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Startup.cs]: fixed

File: DigitalOffice/Services/UserService/UserService.csproj [hunk 27]
```diff
@@ -13,4 +13,8 @@
     </PackageReference>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Controllers\" />
```
Comment (KostinP) [inline DigitalOffice/Services/UserService/UserService.csproj:17]: Strange, there is a controller, but it includes a folder.

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs (hunk 16 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Commands/UserCreateCommand.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 24 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 23 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 22 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 21 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 20 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 19 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs (hunk 18 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/CommandsTests/UserCreateCommandTests.cs]: fixed

File: DigitalOffice/Services/UserService/Repositories/UserRepository.cs (hunk 17 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Repositories/UserRepository.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs (hunk 25 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs (hunk 26 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/ValidatorsTests/UserCreateRequestValidatorTests.cs]: fixed

File: DigitalOffice/Services/UserService/UserService.csproj (hunk 27 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/UserService.csproj:17]: Because we will use many controllers (maybe)

Comment (KostinP) [reply DigitalOffice/Services/UserService/UserService.csproj:17]: You misunderstood me. The controller is already in the files, but for some reason it is still added in the configuration file.

File: DigitalOffice/Services/UserService/Startup.cs [hunk 28]
```diff
@@ -25,6 +35,31 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddControllers();
+
+            CreateTransientCommands(services);
+            CreateTransientRepositories(services);
+            CreateTransientValidators(services);
+            CreateTransientMappers(services);
+        }
+
+        private void CreateTransientCommands(IServiceCollection services)
```
Comment (DarkFelas) [inline DigitalOffice/Services/UserService/Startup.cs]: Rename ConfigureCommands
Change other methods as well

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 24 above)
Comment (DarkFelas) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: No, I still see repository.UserCreate(user); here

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 22 above)
Comment (DarkFelas) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Not fixed, where user is added to dbContext?

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+using System.Collections.ObjectModel;
+using System.Data.Entity;
+using System.Security.Cryptography;
+using System.Text;
+using UserService.Database;
+using UserService.Database.Entities;
+using UserService.Repositories;
+using UserService.Repositories.Interfaces;
+
+namespace UserServiceUnitTests.RepositoriesTests
```
Comment (DarkFelas) [inline DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: Rename folder -> Repositories
Change namespace -> UserServiceUnitTests.Repositories

**Review by DarkFelas — CHANGES_REQUESTED** [review]

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 29 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

File: DigitalOffice/Services/UserService/Startup.cs (hunk 28 above)
Comment (denis-kriv) [reply DigitalOffice/Services/UserService/Startup.cs]: fixed

File: DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs (hunk 22 above)
Comment (denis-kriv) [reply DigitalOffice/Tests/UserServiceTests/RepositoriesTests/UserRepositoryTests.cs]: fixed

**Comment by DarkFelas** [comment]
@KostinP what type of changes do you want?

**Review by DarkFelas — APPROVED** [review]

**Review by Inzilbeth — APPROVED** [review]

File: DigitalOffice/Services/UserService/Mappers/UserCreateRequestToDbUserMapper.cs [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace UserService.Mappers
+{
+    public class UserCreateRequestToDbUserMapper : IMapper<UserCreateRequest, DbUser>
+    {
+        public DbUser Map(UserCreateRequest request)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Mappers/UserCreateRequestToDbUserMapper.cs]: Use 'nameof' expression to reference parameter 'request' name

File: DigitalOffice/Services/UserService/Repositories/Interfaces/IUserRepository.cs [hunk 31]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Repositories/Interfaces/IUserRepository.cs]: Using directive is not required by the code and can be safely removed

File: DigitalOffice/Services/UserService/Commands/Interfaces/IUserCreateCommand.cs [hunk 32]
```diff
@@ -0,0 +1,9 @@
+﻿using UserService.RestRequests;
+
+namespace UserService.Interfaces
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Commands/Interfaces/IUserCreateCommand.cs]: Change namespace

File: DigitalOffice/Tests/UserServiceTests/Mappers/UserCreateRequestToDbUserMapperTests.cs [hunk 33]
```diff
@@ -0,0 +1,65 @@
+﻿using NUnit.Framework;
+using System;
+using System.Collections.Generic;
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/UserServiceTests/Mappers/UserCreateRequestToDbUserMapperTests.cs]: Using directive is not required by the code and can be safely remove

**Review by KostinP — APPROVED** [review]

**Pull request merged**